content/blogposts/2023/slow-to-simd.md

```diff
@@ -0,0 +1,508 @@
+[Rust](https://nnethercote.github.io/perf-book/bounds-checks.html).
+
+Exercise for the reader: why is it significant that we slice like `a[i:i+4:i+4]` rather than just `a[i:i+4]`?
+
+## Quantization
+
+We've improved single-core search throughput by ~50% at this point, but now we've hit a new bottleneck: memory usage.
+Our vectors 1536 dimensions. With 4-byte elements, this comes out to 6KiB per vector, and we generate roughly a million
+vectors per GiB of code. That adds up quickly. We had a few customers come to us with some massive monorepos, and we
+wanted to reduce our memory usage so we can support those cases more cheaply.
+
+One possible mitigation would be to move the vectors to disk, but loading them from disk at search time can add
+[significant latency](https://colin-scott.github.io/personal_website/research/interactive_latency.html), especially on
+slow disks. Instead, we chose to compress our vectors with `int8` quantization. 
+
+There are [plenty of ways](https://en.wikipedia.org/wiki/Dimensionality_reduction) to compress vectors, but we'll be
+talking about [_integer quantization_](https://huggingface.co/docs/optimum/concept_guides/quantization), which is
+relatively simple, but effective. The idea is to reduce the precision of the 4-byte `float32` vector elements by
+converting them to 1-byte `int8`s.
+
+I won't get into exactly _how_ we decide to do the translation between `float32` and `int8`, since that's a pretty [deep
+topic](https://huggingface.co/docs/optimum/concept_guides/quantization), but suffice it to say our function now looks
+like the following:
+
+```go
+func DotInt8BCE(a, b []int8) int32 {
+	if len(a) != len(b) {
+		panic("slices must have equal lengths")
+	}
+
+	sum := int32(0)
+	for i := 0; i < len(a); i += 4 {
+		aTmp := a[i : i+4 : i+4]
+		bTmp := b[i : i+4 : i+4]
+		s0 := int32(aTmp[0]) * int32(bTmp[0])
+		s1 := int32(aTmp[1]) * int32(bTmp[1])
+		s2 := int32(aTmp[2]) * int32(bTmp[2])
+		s3 := int32(aTmp[3]) * int32(bTmp[3])
+		sum += s0 + s1 + s2 + s3
+	}
+	return sum
+}
+```
+
+This change yields a 4x reduction in memory usage at the cost of some accuracy (which we carefully measured, but is
+irrelevant to this blog post).
+
+Unfortunately, re-running the benchmarks shows our search speed regressed a bit from the change. Taking a look at the
+generated assembly (with `go tool compile -S`), there are some new instructions for converting `int8` to `int32`, which
+might explain the difference. I didn't dig too deep though, since all our performance improvements up to this point
+become irrelevant in the next section.
+
+<Chart
+    rows={[
+        {name: "DotNaive", value: 941900},
+        {name: "DotUnroll4", value: 1286000},
+        {name: "DotBCE", value: 1407000},
+        {name: "DotInt8BCE", value: 1240000},
+    ]}
+/>
+
+
+## SIMD
+
+I always love an excuse to play with SIMD. And this problem seemed like the perfect nail for that hammer.
```

Comment:
I think the transition here could be clearer. Otherwise it's a little confusing how this relates, like is it layering on another improvement, or replacing the previous ones? You mention that the previous improvements become irrelevant, but don't explain why.

Maybe you could present it as: the improvements so far proved too incremental and sensitive to changes in data format. So we looked into a more dramatic optimization.

Reply:
Done!